### DISCUSSION TITLE
the relative position between script and autoload.php

### DISCUSSION THREAD

**Discussion by rino1** [composer/composer#12083] [General]
I created a package containing only two files:

```
composer.json
script
```

Let's examine them...

The composer.json file

```
{
     "name": "my/package",
     "version": "1"
}
```

The script file

```
#!/usr/bin/php
<?php
require __DIR__ . "/../../autoload.php";
// do something...
echo "Run successfull!!!" . PHP_EOL;
```

Now I would like to install this package.
But how is it installed?
There are many methods

```
composer create-project ... # autonomously
composer require ... # as dependence
composer install ...
composer update ...
COMPOSER_VENDOR_DIR=/tmp/jgj76 composer ... # you can also set up a custom vendor directory
```

But so as not to give us a headache, we will only look at the first, second and last methods.

So let's proceed...

```
++ composer create-project my/package autonomous
blabla...
++ composer require -d dependence my/package
blabla...
```

This structure was therefore created

```
├── autonomous
│   ├── composer.json
│   ├── composer.lock
│   ├── script
│   └── vendor
│       ├── autoload.php
│       └── composer
│           ├── autoload_classmap.php
│           ├── autoload_namespaces.php
│           ├── autoload_psr4.php
│           ├── autoload_real.php
│           ├── autoload_static.php
│           ├── ClassLoader.php
│           ├── installed.json
│           ├── installed.php
│           ├── InstalledVersions.php
│           └── LICENSE
└── dependence
    ├── composer.json
    ├── composer.lock
    └── vendor
        ├── autoload.php
        ├── composer
        │   ├── autoload_classmap.php
        │   ├── autoload_namespaces.php
        │   ├── autoload_psr4.php
        │   ├── autoload_real.php
        │   ├── autoload_static.php
        │   ├── ClassLoader.php
        │   ├── installed.json
        │   ├── installed.php
        │   ├── InstalledVersions.php
        │   └── LICENSE
        └── my
            └── package
                ├── composer.json
                └── script
```

However, we are only interested in observing the files */script and */autoload.php, or rather, the relative position between them.

Specifically

With regard to dependence install method

```
vendor/my/package/script
vendor/autoload.php
```

While with regard to autonomous install method

```
script
vendor/autoload.php
```

But maybe it's better to do some testing, so you understand the concept straight away

```
++ dependence/vendor/my/package/script
Run successfull!!!
++ autonomous/script
PHP Warning:  require(../../autoload.php): Failed to open stream: No such file or directory in autonomous/script on line 4
```

As you can see, with the package installed as a dependency, everything works.
While, with the package installed autonomous, there is something wrong: in the script file, instead of ../../autoload.php, I should have specified vendor/autoload.php

**In practice the problem is precisely this: how can I guess which method the end user will use to install my package?**
Furthermore, for example, if you then wanted to use a [customized vendor dir](https://getcomposer.org/doc/03-cli.md#composer-vendor-dir), with an absurd name (/tmp/jgj76), 99% of the time there would really be hope of guessing.

```
COMPOSER_VENDOR_DIR=/tmp/jgj76 composer ...
```

So how do I deal with this?
How can I keep up with all the "devilry" of the users?

```
//require __DIR__ . "/../../autoload.php"; // OPTION 1
//require __DIR__ . "/vendor/autoload.php"; // OPTION 2
//require "/tmp/jgj76/autoload.php"; // OPTION 3
// other devilry
```

**Comment by Seldaek** [comment]
The common solution is to check either `vendor/autoload.php` (running your package by itself) or `../../autoload.php` (running your package as a dependency). Those are the only two paths possible.

https://github.com/Seldaek/jsonlint/blob/main/bin/jsonlint#L13-L26 is an example of doing this, note that the file is in bin/ so there is a `../` more than you would need.

**Comment by rino1** [comment]
The problem is that there are many ways to set the vendor directory.

https://getcomposer.org/doc/articles/vendor-binaries.md#finding-the-composer-autoloader-from-a-binary
https://stackoverflow.com/a/31518505/11889508
https://stackoverflow.com/a/36701989/11889508
https://stackoverflow.com/a/39854258/11889508
https://getcomposer.org/doc/06-config.md#vendor-dir

So how do I deal with this?
How can I keep up with all the _devilry_?

**Reply by Seldaek** [reply]
The autoload file is also in the vendor dir so either your package is in it too and you know where autoload is, or it's your package running as root and there you also know the location. So i suggest to listen to my post above and stop imagining devils in your computer

**Comment by rino1** [comment]
However (to prevent duplication of vendor directories) inside the root package the vendor directory should not exist

https://github.com/composer/composer/discussions/11403#discussioncomment-5453127

So how do you do it?

**Reply by Seldaek** [reply]
Do this and stop looking for other ways to contradict me? :D Up to you.. just trying to help here. https://github.com/Seldaek/jsonlint/blob/main/bin/jsonlint#L13-L26

**Reply by rino1** [reply]
But I prefer it like this

```
use const PHP_EOL as EOL;

(function () {
    $errBuffer = ""; // non usare in scrittura (vedere $addErr)
    $addErr = function (string $line) use (&$errBuffer) {
        $errBuffer .= sprintf(
            "%s: %s%s",
            __FILE__,
            $line,
            EOL,
        );
    };

    foreach (["../vendor", "../../.."] as $relativeDirName) {
        $filename = sprintf(
            "%s/%s/autoload.php",
            __DIR__,
            $relativeDirName,
        );

        if (file_exists($filename)) {
            return require_once $filename;
        } else {
            $addErr("WARNING: $filename: File di autoloading inesistente");
        }
    }

    $addErr("Nessun file di autoloading esistente");

    fwrite(STDERR, $errBuffer);
    exit(1);
})();
```

**Reply by Seldaek** [reply]
Sure that works too

**Comment by rino1** [comment]
But (when the package will be installed as root (composer install -d my-package) and **later** as a dependency (composer require -d OTHER-EMPLOYEE-PACKAGE my/package)) this creates the problem of duplicate vendor names

https://github.com/composer/composer/discussions/11403#discussion-5013105

```
./vendor/my/sub1
./vendor/my/sub1/vendor/my/sub2
./vendor/my/sub2
```

Let's summarize better for the specific case

```
OTHER-EMPLOYEE-PACKAGE/vendor/my/package/vendor/autoload.php
                       ^^^^^^            ^^^^^^
                        EVN               DVN

EVN: employee vendor name (created from composer require ...)
DVN: dependency vendor name (this is the duplicated name, created from **previous** composer install ...)
```

**Reply by stof** [reply]
why would you be running `composer install` inside a dependency ? that does not make sense.

**Reply by rino1** [reply]
How about continuing the discussion in this old thread, where you have not been heard from again, and of which I will give you a couple of links

https://github.com/composer/composer/discussions/11403#discussioncomment-9245154
https://github.com/composer/composer/discussions/11403#discussioncomment-9115041

If necessary, reread the thread from the beginning, and respond not immediately, but calmly.

PS

Don't "volatilize" yourself again :)

**Reply by rino1** [reply]
We vanished again :D
